using_restricted_array.rb

```diff
@@ -45,53 +48,61 @@ def search(array, length, value_to_find)
 def find_largest(array, length)
   i = 0
   largest = 0
-  while i < length
+  until array[i] == nil
     if array[i] > largest
       largest = array[i]
     end
     i += 1
   end
+  return largest
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
   i = 0
-  smallest = 0
-  while i < length
-    if array[i] < largest
+  smallest = array[i]
+  until array[i] == nil
+    if array[i] < smallest
       smallest = array[i]
     end
     i += 1
   end
+  return smallest
 end
 
 # Reverses the values in the integer array in place
 def reverse(array, length)
   left = 0
   right = length - 1
-  dummy = 0
+  holder = 0
   while left < length/2
-    dummy = array[left]
+    holder = array[left]
     array[left] = array[right]
-    array[right] = dummy
+    array[right] = holder
     left += 1
     right -= 1
   end
+  return array
```

Comment:
Just to call out: you don't need to return the array since the array is reversed in place and arrays are passed by reference.